### PR TITLE
Make hash more robust to future changes in julia type-printing.

### PR DESCRIPTION
**Pull request by haberdashPI**
## Description

StableHashTraits relies on the string representation of types which has proven to be a fragile assumption (c.f. #43).

This PR provides a solution more robust to julia internals, relying mostly on the assumption that type parameters can be acessed programmatically. This seems like something much less likely to go away (though *how* they're accessed might change, that should be easy to make a bugfix for, if/when that happens).

It does however require creating a new stable_hash `version` and implementing it there. Eventually, (in a separate PR) StableHashTraits should release a breaking version that completely removes version=1 and version=2, since these cannot be trusted to stay stable for Julia 1.11 and later.

In the process of testing out the edge cases for this I found some hash collisions present in version=2 (user defined structs that inherit from `Base.Function` did not hash the fields of said struct, just the type), so I've added a test for this, and removed these collisions for version>2.

## Benchmarks

Author, please:

Run the benchmarks under `test/benchmarks.jl` locally, and update `test/benchmark_records.md` by inserting a new table at the end of the document with the updated numbers
printed at the end of the benchmark script. Place the table as it occurs before
and after the application of this PR.

### CODE REVIEW

File: src/StableHashTraits.jl [hunk 1]
```diff
@@ -425,16 +425,59 @@ function validate_name(str)
...
+qualified_type1_(fn::Function) = qname_(fn, string)
 qualified_name_(x::T) where {T} = qname_(T <: DataType ? x : T, nameof)
-qualified_type_(x::T) where {T} = qname_(T <: DataType ? x : T, string)
+qualified_type1_(x::T) where {T} = qname_(T <: DataType ? x : T, string)
 qualified_(T, ::Val{:name}) = qualified_name_(T)
-qualified_(T, ::Val{:type}) = qualified_type_(T)
+qualified_(T, ::Val{:type1}) = qualified_type1_(T)
+qualified_(T, ::Val{:type2}) = qualified_type2_(T)
+
+"""
```
Comment (haberdashPI) [inline src/StableHashTraits.jl:437]: This function was based of the discussion in #41. The older implementation just replaced "workspace#" strings, but now I can properly make a check rather than just assuming any use of "workspace#" means it's a Pluto variable/module.

File: src/StableHashTraits.jl [hunk 2]
```diff
@@ -22,9 +22,9 @@ these fallback methods will not change even if new fallbacks are defined.
 """
 struct HashVersion{V}
     function HashVersion{V}() where {V}
-        V == 1 && Base.depwarn("HashVersion{1} is deprecated, favor `HashVersion{2}` in " *
-                               "all cases where backwards compatible hash values are not " *
-                               "required.", :HashVersion)
+        V <= 3 && Base.depwarn("HashVersion{V} for V < 3 is deprecated, favor " *
```
Comment (ericphanson) [inline src/StableHashTraits.jl:25]:
```suggestion
        V < 3 && Base.depwarn("HashVersion{V} for V < 3 is deprecated, favor " *
```

File: src/StableHashTraits.jl [hunk 3]
```diff
@@ -495,35 +538,58 @@ julia> stable_typename_id(["a", "b"])
...
-    Likewise, the type names of AbstractArray types are made uniform
-    as their printing changes from Julia 1.6 -> 1.7.
+!!! note If the module of a type is `Core` it is renamed to `Base` before hashing because
+    the location of some types changes between `Core` to `Base` across julia versions.
+    Likewise, the type names of AbstractArray types are made uniform as their printing
+    changes from Julia 1.6 -> 1.7.
+
+!!! warning Version 1 of of `stable_type_id` is deprecated, and its use strongly
+    discouraged. Version 2 relies on an approach that is more robust to future changes in
+    julia internals.
```
Comment (ericphanson) [inline src/StableHashTraits.jl:579]: I thought v3 was the new one?

**Review by ericphanson — COMMENTED** [review]
there are a lot of generated functions and Base internals and various hacks... I wonder if the scope of this package is too broad, and hashing everything is too hard to keep working and stable.

That being said, I get that's the world this package lives in now. Looking at this PR specifically, it looks like line 470 is uncovered, so we probably need a test to hit that branch.

Edge cases seem like custom `<: Function` structs, though this does seem to be working on this branch:
```julia
julia> struct A{T} <: Function
       x::Int
       end

julia> using StableHashTraits

julia> bytes2hex(stable_hash(A{Float64}(1); version=3))
"4c4923c023e60d3fcfbb9f8a62dd37df879a14d7a15ffa2ea8bb040eca4811cb"

julia> bytes2hex(stable_hash(A{Float64}(2); version=3))
"641dc5715937edbb1e525191abad45ea6d45df567547fe499bedc3e71d9c7b20"

julia> bytes2hex(stable_hash(A{Int}(2); version=3))
"60ade5873e7220903baa67749a0f315d640c467b6f26258dea84a4071dc15a8f"

```
Maybe worth a test?

**Comment by haberdashPI** [comment]
> there are a lot of generated functions and Base internals and various hacks... I wonder if the scope of this package is too broad, and hashing everything is too hard to keep working and stable.

I do wonder if the ambition of the v1.0 release (which really expanded how many things were covered by default) has gotten away from me!

This PR aims to reduce, rather than increases, the hacks required though.

I think it would be hard to remove the more broadly scoped hash at this point but maybe it would be good to add something like `SafeHashVersion{1}` that does not dispatch on `Any` (or `T where {T}`) and requires the user to explicitly define `hash_method` for new types.

I'll add the tests.

**Comment by ericphanson** [comment]
> I do wonder if the ambition of the v1.0 release (which really expanded how many things were covered by default) has gotten away from me!

I filed #50 for more thoughts on this; I think the package could be useful while being significantly more scoped down!

**Comment by haberdashPI** [comment]
Superseded by #55
